bin/find-fixable-orphans.py

```diff
@@ -0,0 +1,293 @@
+    )
+    this_all_keys_dict = {}
+
+    # Find S3 keys that are present in NZSL, or absent
+    for video_key in this_s3_bucket_raw_keys_list:
+        dict_row = this_nzsl_raw_keys_dict.get(video_key, None)
+        if dict_row:
+            # NZSL glossvideo record for this S3 key
+            this_all_keys_dict[video_key] = [
+                True,  # NZSL PRESENT
+                True,  # S3 PRESENT
+            ] + dict_row
+        else:
+            # S3 key with no corresponding NZSL glossvideo record
+            this_all_keys_dict[video_key] = [
+                False,  # NZSL Absent
+                True,  # S3 PRESENT
+            ] + [""] * 6
+
+    # Find NZSL keys that are absent from S3 (present in both handled above)
+    for video_key, dict_row in this_nzsl_raw_keys_dict.items():
+        if video_key not in this_s3_bucket_raw_keys_list:
+            # gloss/glossvideo record with no corresponding S3 key
+            # Either:
+            # video_key is real, but the S3 object is missing
+            # video_key is fake (to handle the FULL JOIN) and this gloss/glossvideo never had an S3 object
+            this_all_keys_dict[video_key] = [
+                True,  # NZSL PRESENT
+                False,  # S3 Absent
+            ] + dict_row
+
+    return this_all_keys_dict
+
+
+def find_orphans():
+    all_keys_dict = create_all_keys_dict(
+        get_nzsl_raw_keys_dict(), get_s3_bucket_raw_keys_list()
+    )
+    print("Finding fixable orphans", file=sys.stderr)
+
+    print(CSV_DELIMITER.join(GLOBAL_COLUMN_HEADINGS))
+
+    # Traverse all the NZSL Signbank glosses that are missing S3 objects
+    for video_key, [
+        key_in_nzsl,
+        key_in_s3,
+        gloss_idgloss,
+        gloss_created_at,
+        gloss_id,
+        video_id,
+        gloss_public,
+        video_public,
+    ] in all_keys_dict.items():
+
+        if not key_in_nzsl:
+            # This is an S3 object, not a Signbank record
+            continue
+
+        if key_in_s3:
+            # This Signbank record already has an S3 object, all is well
+            continue
+
+        # Business rule
+        if int(gloss_id) < 8000:
+            continue
+
+        # The gloss_id is the only reliable retrieval key at the Signbank end
+        gloss = Gloss.objects.get(id=gloss_id)
+        gloss_name = gloss.idgloss.split(":")[0].strip()
+        video_path = gloss.get_video_path()
+
+        # Skip any that already have a video path
+        # These should have an S3 object but don't: For some reason the video never made it to S3
+        # These will have to have their videos reinstated (separate operation)
+        if len(video_path) > 0:
+            continue
+
+        # We try to find the orphaned S3 object, if it exists
+        # TODO We could improve on brute-force by installing new libraries eg. rapidfuzz
+        for test_key, [key_nzsl_yes, key_s3_yes, *_] in all_keys_dict.items():
+            if gloss_name in test_key:
```

Comment:
Question: could you skip _fake_ keys here?

Could you start this loop with `if test_key.startswith(FAKEKEY_PREFIX): continue`. There is not much value in  the follow on checks if it is a fake key right? It might even result in false positives against the gloss id.

Reply:
I'll have to wrap my head around the way this logic works again.